test.js

```diff
@@ -178,6 +288,71 @@ test('should not compress on missing header', t => {
   })
 })
 
+test('should decompress compressed Buffers on missing header', t => {
+  t.plan(4)
+  const fastify = Fastify()
+  fastify.register(compressPlugin, { threshold: 0, inflateIfDeflated: true })
+  const json = { hello: 'world' }
+
+  fastify.get('/', (req, reply) => {
+    reply.send(zlib.gzipSync(JSON.stringify(json)))
+  })
+
+  fastify.inject({
+    url: '/',
+    method: 'GET'
+  }, (err, res) => {
+    t.error(err)
+    t.strictEqual(res.statusCode, 200)
+    t.notOk(res.headers['content-encoding'])
+    t.deepEqual(JSON.parse('' + res.payload), json)
+  })
+})
+
+// This throws an error, because the Reply's `send` method attempts
+// to JSON.parse the compressed string before the `onSend` hook is invoked
+// test('should decompress compressed Strings on missing header', t => {
+//   t.plan(4)
+//   const fastify = Fastify()
+//   fastify.register(compressPlugin, { threshold: 0, inflateIfDeflated: true })
+//   const json = { hello: 'world' }
+
+//   fastify.get('/', (req, reply) => {
+//     reply.send(zlib.gzipSync(JSON.stringify(json)).toString())
+//   })
+
+//   fastify.inject({
+//     url: '/',
+//     method: 'GET'
+//   }, (err, res) => {
+//     t.error(err)
+//     t.strictEqual(res.statusCode, 200)
+//     t.notOk(res.headers['content-encoding'])
+//     t.deepEqual(JSON.parse('' + res.payload), json)
+//   })
+// })
```

Comment:
why is this commented out?

Reply:
Mostly to draw attention, see the comment above:
>
```
// This throws an error, because the Reply's `send` method attempts
// to JSON.parse the compressed string before the `onSend` hook is invoked
```
>
Looking back I think I meant to say JSON.stringify instead of JSON.parse, but either way this test fails because `reply.send()` attempts to serialize the stringified gzipped buffer. It seems pretty edge-casey that anyone would do this, but I added the test for completeness.

Reply:
I'm a bit lost. This test should pass because a buffer is sent as-is: https://github.com/fastify/fastify/blob/4b5f818b08d0eea02c46c782281fe90ff596830a/lib/reply.js#L65-L70.

Reply:
In this test the payload is the utf8 string of the gzipped Buffer:
>
```
zlib.gzipSync(JSON.stringify(json)).toString()
```
>

Reply:
Strings are trated the same. Why is this failing?

Reply:
I dug deeper, and it looked like utf8-encoding the gzipped buffer and decoding only the first 4 characters was producing a Buffer that didn't pass the `isGzip()` check, and falling through to the normal `reply.serialize()` logic. So I removed the auto-inflate of strings when there's no header, since I doubt we want to sniff the encoding and decode every string that gets sent out.